security/elastic.py

```diff
@@ -0,0 +1,106 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import json
+import logging
+
+import elasticsearch
+from elasticsearch import helpers
+
+from security import base
+
+INDEX = "ms_security_"
+
+
+class StorageException(Exception):
+    pass
+
+
+class Backend(object):
+    """Represent elasticsearch backend.
+
+    usage:
+        b = elastic.Backend(urls=["url1", "url2"])
+        issues = b.get_issues("Region1", "IssueType1,IssueType2")
+        b.begin()
+        b.store_issue(Issue(...))
+        b.commit("region1")
+    """
+
+    def __init__(self, urls, **kwargs):
+        self.index = INDEX
+        self.es = elasticsearch.Elasticsearch(urls)
+        self.body = ""
+
+    def get_issues(self, region, types=None, discovered_after=None):
+        query = {
+            "query": {
+                "bool": {
+                    "must_not": {
+                        "exists": {
+                            "field": "resolvedAt",
+                        },
+                    },
+                },
+            }
+        }
+        if discovered_after:
+            query["query"]["bool"]["must"] = {
+                "range": {
+                    "discoveredAt": {"gt": discovered_after.isoformat("T")}
+                }
+            }
+        issues = []
+        try:
+            for i in helpers.scan(self.es, index=self.index + region,
+                                  doc_type=types, query=query, scroll="1m"):
+                kwargs = i["_source"]
+                kwargs["discovered_at"] = kwargs["discoveredAt"]
+                kwargs["confirmed_at"] = kwargs["confirmedAt"]
+                kwargs["resolved_at"] = kwargs.get("resolvedAt")
+                issues.append(base.Issue(i["_type"], **kwargs))
+            return issues
+        except elasticsearch.NotFoundError:
+            return []
+
+    def begin(self):
+        if self.body:
+            logging.warning("Unsaved data %s", self.body)
+        self.body = ""
+
+    def commit(self, region):
+        try:
+            self.es.indices.create(index=self.index + region)
+        except elasticsearch.RequestError as ex:
+            if ex.error != "index_already_exists_exception":
+                raise
+        retval = self.es.bulk(self.body, self.index + region)
+        self.body = ""
+        try:
+            for item in retval["items"]:
+                if 200 > item["index"]["status"] > 201:
+                    raise KeyError
+        except KeyError:
+            raise StorageException(
+                "Failed to create some items: %s" % retval)
+
+        return retval
+
+    def store(self, issue):
+        self.body += json.dumps({"index": {
```

Comment:
if you are performance/tuning addicted you should use here array of strings not append to string every time. 

Basically when you are appending to string you are creating new string with bigger size... 

Way faster method, is to create array, append to it and use "\n".join()

Take a look here https://waymoot.org/home/python_string/